Fix/#6416 Input Validator (also for #6417, #6418, #6419)

spetros-do opened the pull request:
``` markdown
To Improve:
- Input validation logic
- Error-handling
- Debugging & logging
- Return of correct error codes

This fix covers issues: #6416, #6417, #6418, #6419

maxdiebold-erg commented on src/validators/db-lookup.validator.ts:
```diff
@@ -19,28 +19,53 @@ export class DbLookupValidator implements ValidatorConstraintInterface {
+    // Handle empty or null values
     if (value || !ignoreEmpty) {
+      // Additional check for non-numeric input (assuming IDs are numeric)
+      if (isNaN(value)) {
+        throw new BadRequestException(`Invalid input: "${value}" is not a valid numeric identifier.`);
+      }
```
This error is incorrectly thrown when a string value is passed in for validation. If this validation is desired, it should be applied only to inputs that are _supposed_ to be a number.

spetros-do replied:
``` markdown
Thanks, MaxD! I have made the following changes on < db-lookup.validator.ts> to correct the error you mentioned. 
I have added optional numeric validation with `validateNumeric` flag, improved the error handling and added proper type checking.

spetros-do replied:
``` markdown
On the <.idea/ files>, I was an oversight as I rushing to push before our meeting. Thx.

maxdiebold-erg reviewed (changes requested):
I noted one bug below that is preventing further testing.

Did you intend to include the `.idea/` files in the commits? Or should `.idea/` be added to the root `.gitignore`?

maxdiebold-erg commented on src/pipes/db-lookup.pipe.ts:
```diff
@@ -4,11 +4,18 @@ import {
+    | "primary"
+    | {
+        validateNumeric?: boolean;
+        ignoreEmpty?: boolean;
```
Why is this duplicated here? If it's to provide a keyed object for providing arguments, I think it would be best to refactor the function's parameters to _all_ be in an object. The current way works but I think it could be confusing.

spetros-do replied:
``` markdown
Good catch!
Done in the revised version. 
It now accepts a single configuration object, making the API cleaner and removing any parameter duplication.

maxdiebold-erg commented on src/validators/db-lookup.validator.ts:
```diff
@@ -17,55 +17,45 @@ export class DbLookupValidator implements ValidatorConstraintInterface {
       const options =
         findOption === 'primary'
           ? {
-              // Find by the entity's primary key
```
Could just be me, but I think this comment is helpful.

spetros-do replied:
``` markdown
Agreed. Updated the validation logic on the revised version.

maxdiebold-erg commented on src/pipes/db-lookup.pipe.ts:
```diff
@@ -18,7 +25,21 @@ export function DbLookup(
       options: validationOptions,
-      constraints: [{ ignoreEmpty, type, findOption }],
+      constraints: [
+        typeof findOption === "object" && !Array.isArray(findOption)
```
It doesn't look like any type in the `findOption` union type is an Array, you could remove that check.

spetros-do replied:
``` markdown
Agreed & done. 
Removed the unnecessary <!Array.isArray(findOption)>

maxdiebold-erg reviewed (commented):
I included a couple more questions, they're not necessarily requests for changes.

I think including the `validateNumeric` parameter solves the issue I'd noted previously, but it will push the work of configuration to the front-ends (which I know we were trying to avoid). This is fine if we want to do this, but then we'll need to update the other pipes that utilize the `DbLookupValidator`, too, so they can pass in that parameter.

spetros-do commented:
```markdown
**Files Updated:** 
1. `src/validators/db-lookup.validator.ts`:
   - Added environment-aware error handling
   - Enhanced error messages
   - Maintained helpful comments about PK resolution
   - Added proper error transformation
2. `src/pipes/db-lookup.pipe.ts`:
   - Removed unnecessary array check
   - Simplified configuration structure
   - Added `DbLookupDecoratorConfig` interface
   - Improved type safety
3. `src/interfaces/validator-options.interface.ts`:
   - Added `validateNumeric` option
   - Enhanced type definitions
   - Maintained backward compatibility
4. Added `src/validators/db-lookup.validator.spec.ts`:
   - Comprehensive test coverage
   - Error handling scenarios
   - Validation test cases
   - Environment-aware testing

**Acceptance Criteria:**
- [ ] Validator correctly handles numeric and string inputs
- [ ] Error messages are clear and user-friendly
- [ ] Test coverage is comprehensive (>95%)
- [ ] No console errors in test environment
- [ ] Maintains backward compatibility
- [ ] Build succeeds without warnings
- [ ] All tests pass successfully

 **Testing to be Completed:** 
- [ ] Numeric validation 
- [ ] String validation
- [ ] Empty value handling
- [ ] Primary key lookups
- [ ] Custom finder functionality
- [ ] Error scenarios
- [ ] Environment-aware logging

maxdiebold-erg commented on src/pipes/db-lookup.pipe.ts:
```diff
@@ -4,42 +4,31 @@ import {
-import { DbLookupValidator } from '../validators';
+import { DbLookupValidator } from "../validators/db-lookup.validator";
+
+export interface DbLookupConfig {
+  type: any;
+  findOption?: ((validationArguments: ValidationArguments) => FindOneOptions) | "primary";
+  validateNumeric?: boolean;
+  ignoreEmpty?: boolean;
+}
```
Unfortunately, moving these all into an object is _not_ backwards compatible. I think the least intrusive way would be to just use the last option for `DbLookup`-related options:
```suggestion
export interface DbLookupConfig {
  validateNumeric?: boolean;
  ignoreEmpty?: boolean;
}
```

spetros-do replied:
```markdown
Pls see [https://github.com/US-EPA-CAMD/easey-common/pull/175](url)

maxdiebold-erg commented on src/pipes/db-lookup.pipe.ts:
```diff
@@ -4,42 +4,31 @@ import {
-      },
-  validationOptions?: ValidationOptions,
-  ignoreEmpty = true
+  config: DbLookupConfig,
+  validationOptions?: ValidationOptions
```
Related to the comment above, this breaks the current interface for the pipe. The `ignoreEmpty` parameter, however, is not used in any place the pipe is used, so I think we could replace that parameter:
```suggestion
  type: any,
  findOption?: ((validationArguments: ValidationArguments) => FindOneOptions) | "primary",
  validationOptions?: ValidationOptions,
  config: DbLookupConfig,
```

spetros-do replied:
Pls see PR #170.

maxdiebold-erg commented on src/pipes/db-lookup.pipe.ts:
```diff
@@ -4,42 +4,31 @@ import {
-              findOption,
-            },
-      ],
+      constraints: [{
+        ignoreEmpty: config.ignoreEmpty ?? true,
+        validateNumeric: config.validateNumeric ?? false,
+        type: config.type,
+        findOption: config.findOption
+      }],
```
_If_ the approach noted above is used, this would need to change to:
```suggestion
      constraints: [{
        ignoreEmpty: config.ignoreEmpty ?? true,
        validateNumeric: config.validateNumeric ?? false,
        type,
        findOption,
      }],
```

spetros-do replied:
Pls see PR #170.

maxdiebold-erg reviewed (commented):
Good call on adding the unit tests!

I like the de-duplication here, but unfortunately this will break current calls with this pipe. Here's an example of a current call:

```typescript
  @DbLookup(
    Plant,
    (args: ValidationArguments): FindOneOptions<Plant> => {
      return { where: { orisCode: args.value } };
    },
    {
      message: (args: ValidationArguments) => {
        return CheckCatalogService.formatResultMessage('IMPORT-24-A', {
          orisCode: args.value,
        });
      },
    },
  )
```

The call signatures are not compatible. If we want to go this route, we will need to update the invocations of the `DbLookup` **pipe** (other pipes that use the `DbLookup` **validator** are fine): I only see a handful of instances of using the pipe directly across a few APIs, and we'll need to update them anyway to add `validateNumeric: true`, since all the calls I see are for validating an ORIS code.

The suggestions below are for an alternative solution, where only the last parameter is changed (which is currently not used anywhere). I think either solution is fine, but if these most recent changes are adopted then the changes in the calling APIs will need to be made simultaneously to avoid type errors.

Also, I'm not sure what happened to the `.gitignore`, it's now registering as a binary file?

spetros-do commented on src/validators/is-valid-codes.validator.ts:
```diff
@@ -2,37 +2,61 @@ import {
+      } catch (error) {
+        // Log the error and throw a meaningful exception
+        console.error('Error during code validation:', error);
+        throw new BadRequestException('Internal server error during code validation.');
```
@maheese Understood. Will make the update accordingly. Thanks.

spetros-do replied:
Pls see PR #170.

spetros-do commented:
see PR# 175
